From: Hendrik Brueckner <brueckner@redhat.com>
Date: Tue, 13 Apr 2010 11:43:10 -0400
Subject: [s390] nss: add missing .previous call to asm function
Message-id: <20100413114310.GC21670@redhat.com>
Patchwork-id: 24121
O-Subject: [RHEL5.6 PATCH 1/1] [s390] nss: add missing .previous statement to
	asm function
Bugzilla: 581522
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

Description
-----------
The savesys_ipl_nss() asm function is put into the .init.text section.
However, it is missing a ".previous" statement which would restore the
previous section.  Depending on the compiler version used to built the
kernel, code may or may not be placed into the correct section. If the
code is put into the wrong section, the kernel might panic or fail to
IPL.

To solve this issue, add a the ".previous" statement to return to the
previous section.

Bugzilla
--------
BZ 581522
https://bugzilla.redhat.com/show_bug.cgi?id=581522

Upstream status of the patch
----------------------------
The patch will be upstream as of kernel version 2.6.34
http://git.kernel.org/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commitdiff;h=187644636ea79aa4bdc04757ebca919c4fd546e0

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.


diff --git a/arch/s390/kernel/setup.c b/arch/s390/kernel/setup.c
index c24457c..6152d6c 100644
--- a/arch/s390/kernel/setup.c
+++ b/arch/s390/kernel/setup.c
@@ -287,7 +287,8 @@ asm(
 	"	lm	6,15,24(15)\n"
 #endif
 	"	br	14\n"
-	"	.size	savesys_ipl_nss, .-savesys_ipl_nss\n");
+	"	.size	savesys_ipl_nss, .-savesys_ipl_nss\n"
+	"	.previous\n");
 
 static __init void create_kernel_nss(void)
 {
